### PR TITLE
Add double write policy to allow same content double writes

### PR DESCRIPTION
**Pull request by smera**
Add a double write policy that allows for double writes as long as the written content is the same.

### CODE REVIEW

File: Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs [hunk 1]
```diff
@@ -17,6 +17,8 @@
 using BuildXL.Utilities.Tracing;
 using JetBrains.Annotations;
 using static BuildXL.Utilities.FormattableStringEx;
+using BuildXL.Storage;
+using System.Collections.ObjectModel;
```
Comment (erikma) [inline Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs]: nit: Odd ordering of these new usings, should shuffle into existin

File: Public/Src/Engine/Scheduler/IFileMonitoringViolationAnalyzer.cs [hunk 2]
```diff
@@ -25,14 +27,21 @@ public interface IFileMonitoringViolationAnalyzer
         /// Analyzes an unordered sequence of violations for a single pip. This may be called only once per pip.
         /// </summary>
         /// <returns>true if there were no violations marked as error</returns>
+        /// <remarks>
+        /// If double writes involving same-content files are detected but allowed, a collection of those non-reported violations 
+        /// are returned. This is usefule for the case of cache convergence, where outputs may change after the main violation analysis
```
Comment (erikma) [inline Public/Src/Engine/Scheduler/IFileMonitoringViolationAnalyzer.cs]: >usefule [](start = 34, length = 7)

useful

File: Public/Src/Engine/UnitTests/Scheduler/FileMonitoringViolationAnalyzerTests.cs [hunk 3]
```diff
@@ -125,7 +132,9 @@ public void DoubleWriteEventLoggedOnViolationError()
                 exclusiveOpaqueDirectoryContent: null,
                 sharedOpaqueDirectoryWriteAccesses: null,
                 allowedUndeclaredReads: null,
-                absentPathProbesUnderOutputDirectories: null);
+                absentPathProbesUnderOutputDirectories: null,
+                ReadOnlyArray<(FileArtifact, FileMaterializationInfo, PipOutputOrigin)>.Empty,
+                out var _);
```
Comment (erikma) [inline Public/Src/Engine/UnitTests/Scheduler/FileMonitoringViolationAnalyzerTests.cs]: >out var _ [](start = 16, length = 9)

nit: Can just be 'out _' (many of these)

File: Public/Src/Engine/UnitTests/Scheduler/PipQueueTest.cs [hunk 4]
```diff
@@ -651,6 +651,7 @@ public class TestPipExecutor
                                 pipScope,
                                 executionResult,
                                 cacheableProcess.Process,
+                                out var _,
                                 out var _);
```
Comment (erikma) [inline Public/Src/Engine/UnitTests/Scheduler/PipQueueTest.cs]: out _

File: Public/Src/Engine/UnitTests/Scheduler/FileMonitoringViolationAnalyzerTests.cs (hunk 3 above)
Comment (smera) [reply Public/Src/Engine/UnitTests/Scheduler/FileMonitoringViolationAnalyzerTests.cs]: Yeap. Keep forgetting about that.

---
In reply to: [279903344](https://github.com/Microsoft/BuildXL/pull/224#discussion_r279903344) [](ancestors = 279903344)

File: Public/Src/Utilities/Configuration/DoubleWritePolicy.cs [hunk 5]
```diff
@@ -32,16 +34,34 @@ public enum DoubleWritePolicy : byte
...
+                    return true;
+                default:
+                    throw new InvalidOperationException("Unexpected double write policy " + policy.ToString());
+            }
+        }
+
+        /// <summary>
+        /// Whether the double-write policy implies that double writes are possible without implying a build break
+        /// </summary>
+        public static bool ImpliesDoubleWriteCanHappen(this DoubleWritePolicy policy)
```
Comment (ljcollins25) [inline Public/Src/Utilities/Configuration/DoubleWritePolicy.cs]: >CanHappen [](start = 45, length = 9)

Allowed

File: Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs [hunk 6]
```diff
@@ -575,78 +662,137 @@ private Pip TryFindProducer(AbsolutePath producedPath, VersionDisposition versio
...
-                                related: relatedPip,
-                                // we don't have the path of the process that caused the file access violation, so 'blame' the main process (i.e., the current pip) instead
-                                pip.Executable.Path));
-                    }
+        private void ReportWriteViolations(
+            Process pip, 
+            List<ReportedViolation> reportedViolations, 
+            IReadOnlyDictionary<FileArtifact, FileMaterializationInfo> outputArtifactsInfo, 
+            AbsolutePath access,
+            [CanBeNull] Dictionary<FileArtifact, (FileMaterializationInfo fileMaterializationInfo, ReportedViolation reportedViolation)> allowedDoubleWriteViolations)
```
Comment (dannyvv) [inline Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs]: >Dictionary<FileArtifact, (FileMaterializationInfo fileMaterializationInfo, ReportedViolation reportedViolation)> [](start = 24, length = 112)

This is typed so many times, maybe it is worth considering adding a type alias.

File: Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs [hunk 7]
```diff
@@ -73,8 +75,9 @@ private enum DynamicFileAccessType : byte
...
         private readonly IExecutionLogTarget m_executionLog;
         private readonly ConcurrentBigMap<AbsolutePath, UndeclaredAccessors> m_undeclaredAccessors = new ConcurrentBigMap<AbsolutePath, UndeclaredAccessors>();
-        // Maps of paths that are dynamically read or written to the corresponding pip. The tuple indicates whether it is a reader of the path (or a writer), together with the actual pip
-        private readonly ConcurrentBigMap<AbsolutePath, (DynamicFileAccessType accessType, Process pip)> m_dynamicReadersAndWriters = new ConcurrentBigMap<AbsolutePath, (DynamicFileAccessType, Process)>();
+        // Maps of paths that are dynamically read or written to the corresponding pip. 
+        // The tuple indicates whether it is a reader of the path or a writer, together with the actual pip and materialization info when available (absent file otherwise)
+        private readonly ConcurrentBigMap<AbsolutePath, (DynamicFileAccessType accessType, Process pip, FileMaterializationInfo fileMaterializationInfo)> m_dynamicReadersAndWriters = new ConcurrentBigMap<AbsolutePath, (DynamicFileAccessType, Process, FileMaterializationInfo)>();
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs]: >Process [](start = 91, length = 7)

Make this a pip id.

File: Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs (hunk 6 above)
Comment (smera) [reply Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs]: Adding aliases to (named) value tuples is only available in >= C#7.0...

File: Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs [hunk 8]
```diff
@@ -656,7 +802,12 @@ private Pip TryFindProducer(AbsolutePath producedPath, VersionDisposition versio
             {
                 foreach (FileArtifact fileArtifact in directoryContent)
                 {
-                    RegisterWriteInPathAndUpdateViolations(pip, fileArtifact.Path, reportedViolations);
+                    var outputArtifactInfo = GetOutputMaterializationInfo(pip, outputArtifactsInfo, fileArtifact);
+                    RegisterWriteInPathAndUpdateViolations(pip, fileArtifact.Path, reportedViolations, outputArtifactInfo, out ReportedViolation? allowedSameContentDoubleWriteViolation);
+                    if (allowedSameContentDoubleWriteViolations != null && allowedSameContentDoubleWriteViolation.HasValue)
+                    {
```
Comment (ljcollins25) [inline Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs]: Is this case allowed?

Comment (smera) [reply Public/Src/Engine/Scheduler/FileMonitoringViolationAnalyzer.cs]: Yeap, good call, this is not needed for exclusive opaques.

---
In reply to: [280218947](https://github.com/Microsoft/BuildXL/pull/224#discussion_r280218947) [](ancestors = 280218947)

**Pull request merged**